### ISSUE TITLE
Regular Expressions Are Too Greedy

### ISSUE THREAD

**Issue by Capital-EX** [factor/factor#2679]
It appears that Factors `regexp` might not correctly handle alternatives. When testing the regex `"(\\\\|\\"|[^"])*"`, Regex101 gives the following matches:
![image](https://user-images.githubusercontent.com/13488718/188662595-9a531e4b-761f-4bc0-9e2c-5869ccb66069.png)

However, when testing this in Factor, it gives the following matches:
![image](https://user-images.githubusercontent.com/13488718/188662667-3992ed19-9517-49c8-a02a-7aaeb8c04d2b.png)

It appears that `regepx` will always try to match the most broad pattern. Thus `[^"]` will always take precedence over `\\` which allows for `"\\" "` to be matched instead of the more specific `"\\"`

**Comment by mrjbq7** [comment]
```factor
"\"\\\\\" \"Hello, World!\"  \"\\\"Hello, World!\\\"\""
R/ "(\\\\|\\"|[^"])+"/ all-matching-subseqs [ print nl flush ] each
```

**Comment by Capital-EX** [comment]
Doesn't seem to change anything. Additionally it wouldn't match an empty string:

![image](https://user-images.githubusercontent.com/13488718/188680590-1269bca4-7904-486f-b8f7-c4f476a5a68a.png)

**Comment by mrjbq7** [comment]
Sorry, I just pasted the code to run so we could reference it instead of the screenshot :-)

**Comment by Capital-EX** [comment]
OH, yeah, that's smart, sorry! I'll remember that for next time.

**Comment by error256** [comment]
The only relevant documentation I've been able to find is for [`all-matching-slices`](https://docs.factorcode.org/content/word-all-matching-slices%2Cregexp.html):
> Finds a sequence of disjoint substrings which each match the pattern. It chooses this by finding **the leftmost longest match**, and then the leftmost longest match which starts after the end of the previous match, and so on.

So it seems to work as documented. This behavior isn't very common across engines, but isn't anything new. For example, grep returns the same result with POSIX regexps.

**Comment by Capital-EX** [comment]
Yeah, then that's what tripped me up. I'm used to first-come-first-matched approach to regexps.

**Comment by razetime** [comment]
@Capital-EX does `all-matching-slices` fit the correct description, then?
